flowmachine/tests/test_cache.py

```diff
@@ -249,6 +253,80 @@ def test_retrieve(get_dataframe):
     assert get_dataframe(from_cache[dl1.query_id]).equals(local_df)
 
 
+@pytest.fixture
+def create_and_store_novel_query():
+    """Creates and stores a query object defined at runtime."""
+
+    class TestQuery(Query):
+        @property
+        def column_names(self):
+            return ["value"]
+
+        def _make_query(self):
+            return "select 1 as value"
+
+    q = TestQuery()
+    q_id = q.query_id
+    q.store().result()
+    yield q_id
+
+
+def test_retrieve_novel_query(create_and_store_novel_query, flowmachine_connect):
+    """Test that a runtime defined query can be pulled from cache and invalidated."""
+    from_cache = get_obj_or_stub(get_db(), create_and_store_novel_query)
+    assert from_cache.query_id == create_and_store_novel_query
+    assert from_cache.is_stored
+    from_cache.invalidate_db_cache()
+    assert not from_cache.is_stored
+
+
+@pytest.fixture
+def create_and_store_novel_query_with_dependency():
+    """Creates and stores a query object defined at runtime with a composed query type."""
+
+    class TestQuery(Query):
+        @property
+        def column_names(self):
+            return ["value"]
+
+        def _make_query(self):
+            return "select 1 as value"
+
+    class NestTestQuery(Query):
+        def __init__(self):
+            self.nested = TestQuery()
+            super().__init__()
+
+        @property
+        def column_names(self):
+            return ["value"]
+
+        def _make_query(self):
+            return "select 1 as value"
+
+    q = NestTestQuery()
+    q_id = q.query_id
+    nested_id = q.nested.query_id
+    q.store(store_dependencies=True).result()
+    Query._QueryPool.clear()
+    yield q_id, nested_id
+
```

Comment:
Only extra test I can think of here is nesting a novel in a not-novel or vice versa

Reply:
Wait, novel in not-novel would never happen (if I've understood correctly) and from what I've seen, nesting two novel queries encapsulates nesting a not-novel in a novel, so disregard.